src/session.rs

```diff
@@ -97,20 +91,45 @@ impl Keypair {
         // input keying material
         let ikm = common_secret_padded;
         let salt = None;
-        let info = [];
 
         // output keying material
         let mut okm = [0; 16];
+        hkdf(ikm, salt, &mut okm);
 
-        let (_, hk) = Hkdf::<Sha256>::extract(salt, &ikm);
-        hk.expand(&info, &mut okm)
-            .expect("hkdf expand should never fail");
+        GenericArray::clone_from_slice(&okm)
+    }
+}
 
-        GenericArray::clone_from_slice(okm.as_slice())
+#[cfg(feature = "openssl")]
+fn hkdf(ikm: Vec<u8>, salt: Option<&[u8]>, okm: &mut [u8]) {
+    let mut ctx = openssl::pkey_ctx::PkeyCtx::new_id(openssl::pkey::Id::HKDF)
+        .expect("hkdf context should not fail");
+    ctx.derive_init().expect("hkdf derive init should not fail");
+    ctx.set_hkdf_md(openssl::md::Md::sha256())
+        .expect("hkdf set md should not fail");
+
+    ctx.set_hkdf_key(&ikm)
+        .expect("hkdf set key should not fail");
+    if let Some(salt) = salt {
+        ctx.set_hkdf_salt(salt)
+            .expect("hkdf set salt should not fail");
```

Comment:
I'm not very familiar with the OpenSSL API. Are these failure possibilities present due to possible API misuse only? Allocation errors?

Reply:
Yes, I believe so. For example, calling this method before calling `derive_init()` will lead to an error.